worker/buildbot_worker/util/lineboundaries.py

```diff
@@ -23,56 +23,100 @@
+        # - line times: times when first line symbol was received
+        had_partial_line = False
         if self.partial_line:
-            if len(self.partial_line) > self.MAX_LINELENGTH:
-                if not self.warned:
-                    # Unfortunately we cannot give more hint as per which log that is
-                    log.warn("Splitting long line: {line_start} {length} "
-                             "(not warning anymore for this log)",
-                             line_start=self.partial_line[:30],
-                             length=len(self.partial_line))
-                    self.warned = True
-                # switch the variables, and return previous _partial_line_,
-                # split every MAX_LINELENGTH plus a trailing \n
-                self.partial_line, text = text, self.partial_line
-                ret = []
-                while len(text) > self.MAX_LINELENGTH:
-                    ret.append(text[:self.MAX_LINELENGTH])
-                    text = text[self.MAX_LINELENGTH:]
-                ret.append(text)
-                result = ("\n".join(ret) + "\n")
-                return result
+            had_partial_line = True
             text = self.partial_line + text
-            self.partial_line = None
+            time_partial_line = self.time
+
         text = self.newline_re.sub('\n', text)
-        if text:
-            if text[-1] != '\n':
-                i = text.rfind('\n')
-                if i >= 0:
-                    i = i + 1
-                    text, self.partial_line = text[:i], text[i:]
-                else:
-                    self.partial_line = text
-                    return None
-            return text
-        return None
+
+        lf_positions = self.get_lf_positions(text)
+
+        ret_lines = []  # lines with appropriate number of symbols and their separators \n
+        ret_indexes = []  # ret_indexes is a list of '\n' symbols
+        ret_text_length = -1
+        ret_line_count = 0
+
+        first_position = 0
+        for position in lf_positions:
+            # finds too long lines and splits them, each element in ret_lines will be a line of
+            # appropriate length
+            while position - first_position >= self.max_line_length:
+                line = text[first_position: self.max_line_length - 1] + '\n'
+                ret_lines.append(line)
+                ret_line_count += 1
+                ret_text_length = ret_text_length + len(line)
+                ret_indexes.append(ret_text_length)
+                first_position = first_position + self.max_line_length
+
+            line = text[first_position: (position + 1)]
+            ret_lines.append(line)
+            ret_line_count += 1
+            ret_text_length = ret_text_length + len(line)
+            ret_indexes.append(ret_text_length)
+            first_position = position + 1
+
+        position = len(text)
+        while position - first_position >= self.max_line_length:
```

Comment:
I am worried about the complexification of this algorithm.
I can't see exactly why the algorithm was rewritten.

text management in python must be treated very carefully in python as python is not very performant.

Here the algorithm seems much less performant, and the good reason is not obvious to me. at least it deserves a comment